File: src/hooks/chat/handlers/mealHandlers.ts
import { FunctionCall } from "@google/generative-ai";
import { MealPlan } from "@/data/schema";
import { FunctionHandlerArgs } from "./handlerUtils";

export const handleMealFunctions = async (
  functionCall: FunctionCall,
  args: FunctionHandlerArgs
): Promise<string> => {
  const { addThoughtStep, setPlan } = args;
  let funcResultMsg = "";

  if (functionCall.name === "suggestMeal") {
    addThoughtStep(
      "🔨 Processing meal suggestion",
      "Analyzing preferences and available ingredients to suggest a personalized meal",
      "completed"
    );
    
    const mealArgs = functionCall.args as {
      meal?: {
        name: string;
        calories: number;
        macros: {
          protein: number;
          carbs: number;
          fat: number;
        };
        ingredients: {
          item: string;
          quantity: number;
          unit: string;
        }[];
      };
      justification?: string;
      missing_ingredients?: {
        item: string;
        quantity: number;
        unit: string;
      }[];
    };

    let mealDetails = "Meal suggestion prepared:\n\n";
    
    if (mealArgs.meal) {
      mealDetails += `**Meal:** ${mealArgs.meal.name}\n`;
      mealDetails += `**Estimated Calories:** ${mealArgs.meal.calories}\n`;
      
      if (mealArgs.meal.macros) {
        mealDetails += `**Macros:** ${mealArgs.meal.macros.protein}g protein, ${mealArgs.meal.macros.carbs}g carbs, ${mealArgs.meal.macros.fat}g fat\n`;
      }
      
      if (mealArgs.meal.ingredients && mealArgs.meal.ingredients.length > 0) {
        mealDetails += `\n**Required Ingredients:**\n`;
        mealArgs.meal.ingredients.forEach(ingredient => {
          mealDetails += `- ${ingredient.quantity} ${ingredient.unit} of ${ingredient.item}\n`;
        });
      }
    }
    
    if (mealArgs.missing_ingredients && mealArgs.missing_ingredients.length > 0) {
      mealDetails += `\n**Missing Ingredients (need to buy):**\n`;
      mealArgs.missing_ingredients.forEach(ingredient => {
        mealDetails += `- ${ingredient.quantity} ${ingredient.unit} of ${ingredient.item}\n`;
      });
      mealDetails += "\nAsk the user if they want to add these missing ingredients to their shopping list.\n";
    } else {
      mealDetails += "\n✅ All ingredients are available in your current inventory!\n";
    }
    
    if (mealArgs.justification) {
      mealDetails += `\n**Why this meal was suggested:** ${mealArgs.justification}\n`;
    }
    
    mealDetails += "\nUse this information to present the meal suggestion to the user in a friendly, conversational way.";
    funcResultMsg = mealDetails;
    
    addThoughtStep("✅ Executed: suggestMeal");
  } else if (functionCall.name === "updateMealPlan") {
    const newPlan = functionCall.args as MealPlan;
    setPlan(newPlan);
    funcResultMsg = "Meal plan updated successfully.";
    addThoughtStep("✅ Executed: updateMealPlan");
  }
  return funcResultMsg;
};

File: mise-asi/handlers/meal_handlers.py
"""
Meal Handlers
Maps to: src/hooks/chat/handlers/mealHandlers.ts
"""
from handlers.types import FunctionCall, HandlerContext


def handle_meal_functions(function_call: FunctionCall, ctx: HandlerContext) -> str:
    """
    Handle meal-related function calls
    Maps to: handleMealFunctions in mealHandlers.ts
    """
    name = function_call["name"]
    args = function_call.get("args", {})
    
    if name == "suggestMeal":
        return handle_suggest_meal(args, ctx)
    elif name == "updateMealPlan":
        return handle_update_meal_plan(args, ctx)
    
    return f"Unknown meal function: {name}"


def handle_suggest_meal(args: dict, ctx: HandlerContext) -> str:
    """
    Handle meal suggestion - formats meal data for display
    Note: The LLM generates the meal suggestion, this handler formats it
    """
    try:
        meal = args.get("meal", {})
        justification = args.get("justification", "")
        missing_ingredients = args.get("missing_ingredients", [])
        
        if not meal:
            return "No meal suggestion provided."
        
        # Format meal output
        result = f"**{meal.get('name', 'Meal')}**\n\n"
        result += f"🔥 Calories: {meal.get('calories', 'N/A')}\n"
        
        macros = meal.get("macros", {})
        result += f"💪 Protein: {macros.get('protein', 0)}g | "
        result += f"🍞 Carbs: {macros.get('carbs', 0)}g | "
        result += f"🥑 Fat: {macros.get('fat', 0)}g\n\n"
        
        if meal.get("ingredients"):
            result += "**Ingredients:**\n"
            for ing in meal["ingredients"]:
                result += f"- {ing.get('quantity', '')} {ing.get('unit', '')} {ing.get('item', '')}\n"
            result += "\n"
        
        if justification:
            result += f"💡 *{justification}*\n\n"
        
        if missing_ingredients:
            result += "**Missing ingredients (need to buy):**\n"
            for ing in missing_ingredients:
                result += f"- {ing.get('quantity', '')} {ing.get('unit', '')} {ing.get('item', '')}\n"
        
        ctx.log_step("✅ Executed: suggestMeal")
        return result
        
    except Exception as e:
        ctx.log_step("❌ suggestMeal failed")
        return f"Failed to format meal suggestion: {str(e)}"


def handle_update_meal_plan(args: dict, ctx: HandlerContext) -> str:
    """Update the user's meal plan"""
    try:
        day = args.get("day")
        meal_type = args.get("meal_type")
        meal = args.get("meal", {})
        
        if not day or not meal_type or not meal:
            return "Day, meal type, and meal data are required."
        
        # In frontend, this updates UI state via setPlan
        # Here we just return confirmation
        meal_name = meal.get("name", "meal")
        
        ctx.log_step("✅ Executed: updateMealPlan")
        return f"Updated {day}'s {meal_type} to: {meal_name}"
        
    except Exception as e:
        ctx.log_step("❌ updateMealPlan failed")
        return f"Failed to update meal plan: {str(e)}"
